src/proj.rs

```diff
@@ -0,0 +1,791 @@
+            class: TransformationClass::Affine,
+        }
+    }
+
+    /// Creates an anisotropic scaling (sx,sy).
+    pub fn scale(sx: f32, sy: f32) -> Proj {
+        Proj { 
+            transform: [
+                sx,  0.0, 0.0,
+                0.0, sy,  0.0,
+                0.0, 0.0, 1.0],
+            inverse: [
+                1.0/sx, 0.0, 0.0,
+                0.0, 1.0/sy, 0.0,
+                0.0, 0.0, 1.0],
+            class: TransformationClass::Affine,
+        }
+    }
+
+    /// Inverts the transformation.
+    pub fn inv(self) -> Proj {
+        Proj { transform: self.inverse, inverse: self.transform, class: self.class }
+    }
+
+    /// Calculates a homography from a set of four control point pairs
+    pub fn from_cp(from: [(f32, f32); 4], to: [(f32, f32); 4]) -> Option<Proj> {
+        use rulinalg::matrix::*;
+
+        let (xf1, yf1, xf2, yf2, xf3, yf3, xf4, yf4) = 
+            (from[0].0 as f64, from[0].1 as f64,
+             from[1].0 as f64, from[1].1 as f64,
+             from[2].0 as f64, from[2].1 as f64,
+             from[3].0 as f64, from[3].1 as f64);
+
+        let (x1, y1, x2, y2, x3, y3, x4, y4) =
+            (to[0].0 as f64, to[0].1 as f64,
+             to[1].0 as f64, to[1].1 as f64,
+             to[2].0 as f64, to[2].1 as f64,
+             to[3].0 as f64, to[3].1 as f64);
+
+        let a = Matrix::new(9, 9, vec![
+            0f64,0.0, 0.0,-xf1,-yf1,-1.0, y1*xf1, y1*yf1, y1,
+            xf1, yf1, 1.0, 0.0, 0.0, 0.0,-x1*xf1,-x1*yf1,-x1,
+            0.0, 0.0, 0.0,-xf2,-yf2,-1.0, y2*xf2, y2*yf2, y2,
+            xf2, yf2, 1.0, 0.0, 0.0, 0.0,-x2*xf2,-x2*yf2,-x2,
+            0.0, 0.0, 0.0,-xf3,-yf3,-1.0, y3*xf3, y3*yf3, y3,
+            xf3, yf3, 1.0, 0.0, 0.0, 0.0,-x3*xf3,-x3*yf3,-x3,
+            0.0, 0.0, 0.0,-xf4,-yf4,-1.0, y4*xf4, y4*yf4, y4,
+            xf4, yf4, 1.0, 0.0, 0.0, 0.0,-x4*xf4,-x4*yf4,-x4,
+            xf4, yf4, 1.0, 0.0, 0.0, 0.0,-x4*xf4,-x4*yf4,-x4]);
+                
+        match a.svd() {
+            Err(_) => None,
+            Ok((s, _, v)) => {
+                if s[[8,8]].abs() > 0.0001 || // rank(a) must be 8
+                   s[[7,7]].abs() < 0.0001 {  // but not 7
+                    None
+                } else {
+                    let h = v.col(8).into_matrix().into_vec();
+                    let mut transform = 
+                        [h[0] as f32, h[1] as f32, h[2] as f32,
+                         h[3] as f32, h[4] as f32, h[5] as f32,
+                         h[6] as f32, h[7] as f32, h[8] as f32];
+                    transform = normalize(transform);
+                    let class = TransformationClass::Projection;
+
+                    try_inverse(&transform)
+                        .map(|inverse| { Proj { transform, inverse, class }})
+                }
+            }
+        }
+    }
+
+    // Helper functions used as optiomization in warp
+
+    #[inline(always)]
+    fn warp_p(&self, x: f32, y: f32) -> (f32, f32) {
```

Comment:
Please give these functions clearer names.

Reply:
Renamed to map_projective, map_affine, map_translation